File: src/bmsx/machine/runtime/frame_loop.ts
import { $ } from '../../core/engine';
import type { FrameState, Runtime } from './runtime';
import * as workbenchMode from '../../ide/runtime/workbench_mode';
import { clearBackQueues } from '../../render/shared/queues';
import { clearHardwareLighting } from '../../render/shared/hardware_lighting';
import { RunResult } from '../cpu/cpu';
import { runtimeFault } from '../../ide/runtime/lua_pipeline';

const MAX_FRAME_DELTA = 250;

export class FrameLoopState {
	public currentTimeMs = 0;
	public frameDeltaMs = 0;
	public currentFrameState: FrameState = null;
	public drawFrameState: FrameState = null;

	public reset(): void {
		this.currentTimeMs = 0;
		this.frameDeltaMs = 0;
	}

	public beginFrameState(runtime: Runtime): FrameState {
		if (this.currentFrameState) {
			throw runtimeFault('attempted to begin a new frame while another frame is active.');
		}
		clearHardwareLighting();
		this.frameDeltaMs = runtime.timing.frameDurationMs;
		const budget = runtime.timing.cycleBudgetPerFrame;
		const state: FrameState = {
			haltGame: runtime.debuggerPaused,
			updateExecuted: false,
			luaFaulted: runtime.luaRuntimeFailed,
			cycleBudgetRemaining: budget,
			cycleBudgetGranted: budget,
			cycleCarryGranted: 0,
			activeCpuUsedCycles: 0,
		};
		runtime.machine.vdp.beginFrame();
		runtime.vblank.beginTick();
		this.currentFrameState = state;
		return state;
	}

	public tickUpdate(runtime: Runtime): boolean {
		if (!runtime.tickEnabled) {
			return false;
		}
		runtime.cartBoot.processPending(runtime);
		if (workbenchMode.isOverlayActive(runtime)) {
			if (this.currentFrameState !== null) {
				this.abandonFrameState(runtime);
				return true;
			}
			return false;
		}
		const previousState = this.currentFrameState;
		const previousTickSequence = runtime.frameScheduler.lastTickSequence;
		const previous = {
			pendingCall: runtime.pendingCall,
			frameState: previousState,
		};
		if (this.currentFrameState === null) {
			if (!runtime.frameScheduler.startScheduledFrame(runtime)) {
				return false;
			}
		} else if (previousState !== null && previousState.cycleBudgetRemaining <= 0) {
			if (!runtime.frameScheduler.refillFrameBudget(runtime, this.currentFrameState)) {
				return false;
			}
		}
		this.runActiveFrameState(runtime, this.currentFrameState);
		const nextState = this.currentFrameState;
		if (nextState !== previousState) {
			return true;
		}
		if (nextState !== null && nextState.cycleBudgetRemaining !== (previous.frameState?.cycleBudgetRemaining ?? -1)) {
			return true;
		}
		if (runtime.pendingCall !== previous.pendingCall) {
			return true;
		}
		return runtime.frameScheduler.lastTickSequence !== previousTickSequence;
	}

	public abandonFrameState(runtime: Runtime): void {
		this.currentFrameState = null;
		runtime.vblank.abandonTick();
	}

	public runHostFrame(runtime: Runtime, currentTime: number, runReady: boolean): void {
		if (!$.running) {
			return;
		}
		let hostDeltaMs = 0;
		try {
			$.input.pollInput();
			runtime.screen.beginHostFrame(currentTime);
			workbenchMode.tickIdeInput(runtime);
			workbenchMode.tickTerminalInput(runtime);
			hostDeltaMs = Math.min(currentTime - this.currentTimeMs, MAX_FRAME_DELTA);
			this.currentTimeMs = currentTime;

			if ($.paused) {
				runtime.screen.presentPausedFrame(runtime, hostDeltaMs);
			} else {
				runtime.screen.clearPresentation();
				if (runtime.executionOverlayActive) {
					runtime.screen.runOverlay(runtime);
				} else if (runReady) {
					$.deltatime = runtime.timing.frameDurationMs;
					runtime.screen.syncAfterRuntimeUpdate(runtime, runtime.frameScheduler.run(runtime, hostDeltaMs));
				} else {
					runtime.frameScheduler.clearQueuedTime();
				}
				runtime.screen.presentPending(runtime, hostDeltaMs);
			}
		} catch (error) {
			try {
				workbenchMode.handleLuaError(runtime, error);
				this.abandonFrameState(runtime);
				runtime.screen.presentErrorOverlay(runtime, hostDeltaMs);
			} catch {
				console.error(`Error while handling surfaced game error in runtime: ${error}`);
				this.abandonFrameState(runtime);
			}
		}
		runtime.screen.flushDebugReport(currentTime, runtime);
	}

	private runActiveFrameState(runtime: Runtime, state: FrameState): void {
		try {
			if (runtime.pendingCall === 'entry') {
				this.runUpdatePhase(runtime, state);
				runtime.machine.vdp.flushAssetEdits();
				state.updateExecuted = runtime.pendingCall !== 'entry';
			}
			this.finalizeUpdateSlice(runtime, state);
		} catch (error) {
			runtime.pendingCall = null;
			runtime.vblank.clearHaltUntilIrq(runtime);
			try {
				workbenchMode.handleLuaError(runtime, error);
			} finally {
				if (this.currentFrameState !== null) {
					this.abandonFrameState(runtime);
				}
			}
		}
	}

	private finalizeUpdateSlice(runtime: Runtime, frameState: FrameState): void {
		this.currentFrameState = frameState;
		if (runtime.vblank.tickCompleted || runtime.pendingCall !== 'entry') {
			this.abandonFrameState(runtime);
		}
	}

	private runUpdatePhase(runtime: Runtime, state: FrameState): void {
		if (!runtime.cartEntryAvailable) {
			return;
		}
		if (!runtime.luaGate.ready) {
			return;
		}
		if (state.luaFaulted || runtime.luaRuntimeFailed) {
			state.luaFaulted = true;
			return;
		}
		if (state.haltGame) {
			return;
		}
		try {
			while (runtime.pendingCall === 'entry') {
				if (this.runHaltedUntilIrq(runtime, state)) {
					return;
				}
				if (runtime.vblank.consumeBackQueueClearAfterIrqWake()) {
					clearBackQueues();
				}
				if (runtime.cpuExecution.runWithBudget(runtime, state, state.cycleBudgetRemaining) === RunResult.Halted) {
					return;
				}
				if (runtime.machine.cpu.isHaltedUntilIrq()) {
					if (this.runHaltedUntilIrq(runtime, state)) {
						return;
					}
					continue;
				}
				return;
			}
		} catch (error) {
			state.luaFaulted = true;
			runtime.vblank.clearHaltUntilIrq(runtime);
			throw error;
		}
	}

	private runHaltedUntilIrq(runtime: Runtime, state: FrameState): boolean {
		if (!runtime.machine.cpu.isHaltedUntilIrq()) {
			return false;
		}
		return runtime.vblank.runHaltedUntilIrq(runtime, state);
	}
}

File: src/bmsx_cpp/machine/runtime/frame_loop.cpp
#include "machine/runtime/frame_loop.h"
#include "core/engine.h"
#include "input/manager.h"
#include "machine/runtime/cart_boot.h"
#include "machine/runtime/cpu_executor.h"
#include "machine/runtime/runtime.h"
#include "render/shared/queues.h"
#include <algorithm>

namespace bmsx {
namespace {
constexpr double MAX_FRAME_DELTA_MS = 250.0;

inline double to_ms(std::chrono::steady_clock::duration duration) {
	return std::chrono::duration<double, std::milli>(duration).count();
}
}

void FrameLoopState::reset() {
	frameDeltaMs = 0.0;
}

void FrameLoopState::resetFrameState(Runtime& runtime) {
	frameActive = false;
	runtime.vblank.abandonTick();
	runtime.machine().inputController().restoreSampleArmed(false);
	frameState = FrameState{};
	runtime.vblank.clearHaltUntilIrq(runtime);
	runtime.frameScheduler.reset();
	reset();
	runtime.screen.reset();
	runtime.frameScheduler.resetTickTelemetry();
	runtime.vblank.reset(runtime);
}

void FrameLoopState::beginFrameState(Runtime& runtime) {
	frameActive = true;
	runtime.vblank.beginTick();
	frameState = FrameState{};
	frameState.cycleBudgetRemaining = runtime.timing.cycleBudgetPerFrame;
	frameState.cycleBudgetGranted = runtime.timing.cycleBudgetPerFrame;
	frameState.cycleCarryGranted = 0;
	frameDeltaMs = runtime.timing.frameDurationMs;
	runtime.machine().vdp().beginFrame();
	auto key = [&runtime](std::string_view text) {
		return valueString(runtime.machine().cpu().internString(text));
	};
	auto* const gameTable = asTable(runtime.machine().cpu().getGlobalByKey(key("game")));
	auto* const viewportTable = asTable(gameTable->get(key("viewportsize")));
	const auto viewSize = EngineCore::instance().view()->viewportSize;
	viewportTable->set(key("x"), valueNumber(static_cast<double>(viewSize.x)));
	viewportTable->set(key("y"), valueNumber(static_cast<double>(viewSize.y)));
	auto* const viewTable = asTable(gameTable->get(key("view")));
	auto* const view = EngineCore::instance().view();
	viewTable->set(key("crt_postprocessing_enabled"), valueBool(view->crt_postprocessing_enabled));
	viewTable->set(key("enable_noise"), valueBool(view->applyNoise));
	viewTable->set(key("enable_colorbleed"), valueBool(view->applyColorBleed));
	viewTable->set(key("enable_scanlines"), valueBool(view->applyScanlines));
	viewTable->set(key("enable_blur"), valueBool(view->applyBlur));
	viewTable->set(key("enable_glow"), valueBool(view->applyGlow));
	viewTable->set(key("enable_fringing"), valueBool(view->applyFringing));
	viewTable->set(key("enable_aperture"), valueBool(view->applyAperture));
}

bool FrameLoopState::hasActiveTick(const Runtime& runtime) const {
	return frameActive && runtime.m_luaInitialized && runtime.m_tickEnabled && !runtime.m_runtimeFailed;
}

void FrameLoopState::abandonFrameState(Runtime& runtime) {
	frameActive = false;
	runtime.vblank.abandonTick();
}

void FrameLoopState::finalizeUpdateSlice(Runtime& runtime) {
	if (runtime.m_pendingCall == Runtime::PendingCall::Entry && !runtime.vblank.tickCompleted()) {
		return;
	}
	abandonFrameState(runtime);
}

void FrameLoopState::executeUpdateCallback(Runtime& runtime) {
	try {
		while (true) {
			if (runtime.machine().cpu().isHaltedUntilIrq() && runtime.vblank.runHaltedUntilIrq(runtime, frameState)) {
				return;
			}
			if (runtime.vblank.consumeBackQueueClearAfterIrqWake()) {
				RenderQueues::clearBackQueues();
			}
			if (runtime.m_pendingCall != Runtime::PendingCall::Entry) {
				return;
			}
			const RunResult result = runtime.cpuExecution.runWithBudget(runtime, frameState);
			if (runtime.machine().cpu().isHaltedUntilIrq()) {
				if (runtime.vblank.runHaltedUntilIrq(runtime, frameState)) {
					return;
				}
				continue;
			}
			if (result == RunResult::Halted) {
				runtime.m_pendingCall = Runtime::PendingCall::None;
			}
			return;
		}
	} catch (const std::exception& e) {
		runtime.handleLuaError(e.what());
	}
}

bool FrameLoopState::tickUpdate(Runtime& runtime) {
	if (runtime.m_rebootRequested) {
		runtime.m_rebootRequested = false;
		runtime.frameScheduler.clearQueuedTime();
		if (!EngineCore::instance().rebootLoadedRom()) {
			EngineCore::instance().log(LogLevel::Error, "Runtime fault: reboot to bootrom failed.\n");
		}
		return true;
	}
	if (!runtime.m_luaInitialized || !runtime.m_tickEnabled || runtime.m_runtimeFailed) {
		return false;
	}

	runtime.cartBoot.prepareIfNeeded(runtime);
	if (runtime.cartBoot.pollSystemBootRequest(runtime)) {
		return true;
	}
	if (runtime.cartBoot.processPending(runtime)) {
		return true;
	}

	const bool previousFrameActive = frameActive;
	const int previousRemaining = previousFrameActive ? frameState.cycleBudgetRemaining : -1;
	const bool previousPending = runtime.m_pendingCall == Runtime::PendingCall::Entry;
	const i64 previousSequence = runtime.frameScheduler.lastTickSequence;
	const bool startedFrame = !frameActive;
	if (frameActive) {
		if (frameState.cycleBudgetRemaining <= 0 && !runtime.frameScheduler.refillFrameBudget(runtime, frameState)) {
			return false;
		}
	} else {
		if (!runtime.frameScheduler.startScheduledFrame(runtime)) {
			return false;
		}
	}

	if (runtime.m_pendingCall == Runtime::PendingCall::Entry) {
		executeUpdateCallback(runtime);
	}

	if (startedFrame) {
		auto key = [&runtime](std::string_view text) {
			return valueString(runtime.machine().cpu().internString(text));
		};
		auto* const gameTable = asTable(runtime.machine().cpu().getGlobalByKey(key("game")));
		auto* const viewTable = asTable(gameTable->get(key("view")));
		auto* const view = EngineCore::instance().view();
		auto readViewBool = [](Value value, const char* field) -> bool {
			if (!valueIsBool(value)) {
				throw BMSX_RUNTIME_ERROR(std::string("game.view.") + field + " must be boolean.");
			}
			return valueToBool(value);
		};
		view->crt_postprocessing_enabled = readViewBool(viewTable->get(key("crt_postprocessing_enabled")), "crt_postprocessing_enabled");
		view->applyNoise = readViewBool(viewTable->get(key("enable_noise")), "enable_noise");
		view->applyColorBleed = readViewBool(viewTable->get(key("enable_colorbleed")), "enable_colorbleed");
		view->applyScanlines = readViewBool(viewTable->get(key("enable_scanlines")), "enable_scanlines");
		view->applyBlur = readViewBool(viewTable->get(key("enable_blur")), "enable_blur");
		view->applyGlow = readViewBool(viewTable->get(key("enable_glow")), "enable_glow");
		view->applyFringing = readViewBool(viewTable->get(key("enable_fringing")), "enable_fringing");
		view->applyAperture = readViewBool(viewTable->get(key("enable_aperture")), "enable_aperture");
		runtime.m_debugUpdateCountTotal += 1;
	}

		frameState.updateExecuted = runtime.m_pendingCall != Runtime::PendingCall::Entry;
		runtime.machine().vdp().flushAssetEdits();
		finalizeUpdateSlice(runtime);
	const bool nextFrameActive = frameActive;
	if (nextFrameActive != previousFrameActive) {
		return true;
	}
	if (nextFrameActive && frameState.cycleBudgetRemaining != previousRemaining) {
		return true;
	}
		if ((runtime.m_pendingCall == Runtime::PendingCall::Entry) != previousPending) {
			return true;
		}
	return runtime.frameScheduler.lastTickSequence != previousSequence;
}

void FrameLoopState::runHostFrame(Runtime& runtime, f64 deltaTime, bool platformPaused, bool skipRender) {
	EngineCore& engine = EngineCore::instance();
	if (engine.m_state != EngineState::Running && engine.m_state != EngineState::Paused) {
		return;
	}
	try {
		const auto tickStart = std::chrono::steady_clock::now();
		runtime.screen.recordHostFrame();
		engine.m_last_tick_timing.inputMs = 0.0;
		engine.m_last_tick_timing.workbenchModeInputMs = 0.0;
		engine.m_last_tick_timing.runtimeTerminalInputMs = 0.0;
		engine.m_last_tick_timing.runtimeUpdateMs = 0.0;
		engine.m_last_tick_timing.workbenchModeMs = 0.0;
		engine.m_last_tick_timing.runtimeTerminalMs = 0.0;
		engine.m_last_tick_timing.microtaskMs = 0.0;

		const double hostDeltaMs = std::min(deltaTime * 1000.0, MAX_FRAME_DELTA_MS);
		const double hostDeltaSeconds = hostDeltaMs / 1000.0;
		engine.m_delta_time = hostDeltaSeconds;
		engine.m_total_time += hostDeltaSeconds;
		engine.m_frame_count += 1;
		if (hostDeltaSeconds > 0.0) {
			engine.m_fps = 1.0 / hostDeltaSeconds;
		}

			const auto inputStart = std::chrono::steady_clock::now();
			Input::instance().pollInput();
			const auto inputEnd = std::chrono::steady_clock::now();
			engine.m_last_tick_timing.inputMs = to_ms(inputEnd - inputStart);

			runtime.screen.clearPresentation();
			if (!platformPaused) {
				const i64 previousTickSequence = runtime.frameScheduler.lastTickSequence;
				const auto updateStart = std::chrono::steady_clock::now();
				engine.m_delta_time = runtime.timing.frameDurationMs / 1000.0;
				runtime.frameScheduler.run(runtime, hostDeltaMs);
				runtime.screen.syncAfterRuntimeUpdate(runtime, previousTickSequence);
				const auto updateEnd = std::chrono::steady_clock::now();
				engine.m_last_tick_timing.runtimeUpdateMs = to_ms(updateEnd - updateStart);

				const auto terminalStart = std::chrono::steady_clock::now();
				runtime.machine().vdp().flushAssetEdits();
				const auto terminalEnd = std::chrono::steady_clock::now();
				engine.m_last_tick_timing.runtimeTerminalMs = to_ms(terminalEnd - terminalStart);
			}
		engine.m_delta_time = hostDeltaSeconds;

		if (engine.m_platform && engine.m_platform->microtaskQueue()) {
			const auto microtaskStart = std::chrono::steady_clock::now();
			engine.m_platform->microtaskQueue()->flush();
			const auto microtaskEnd = std::chrono::steady_clock::now();
			engine.m_last_tick_timing.microtaskMs = to_ms(microtaskEnd - microtaskStart);
		}

		engine.m_last_tick_timing.totalMs = to_ms(std::chrono::steady_clock::now() - tickStart);
		if (!skipRender) {
			runtime.screen.render(engine, runtime);
		}
	} catch (const std::exception& e) {
		runtime.handleLuaError(e.what());
	}
}

} // namespace bmsx
